### PR TITLE
Refactor intellij plugins

### LINKED ISSUE #516: IntelliJ plugin not compatible with IDEA 2018.3 EAP
Spek plugin (https://plugins.jetbrains.com/plugin/10915-spek-framework) is not compatible with IDEA >= 183 (IDEA 2018.3 EAP). Are there any plans to support new IDEA? It is extremely hard not to be able to run Spek2 tests from IDE.

### PR DESCRIPTION
**Pull request by raniejade**
With `2018.3` Jetbrains decided to make some changes to certain classes that ultimately force us to duplicate code in order to support previous versions.

Code is mostly the same, most changes are re-organisation (moving to it's own class) and unfortunately duplication. 

Here's also a list of changes to the build:
1. merge the AS plugin source into a single project
2. intellij plugin by default builds the `2018.3` plugin, use `-Pij.version=IJ182` to build for `2018.2`.

I've dropped support for Android Studio 3.1, it's easy to bring back - let me know!

resolves #516

### CODE REVIEW

File: spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidConfigurationType.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.spekframework.intellij
+
+import com.intellij.execution.configurations.ConfigurationTypeBase
+import org.jetbrains.kotlin.idea.KotlinIcons
+
+class SpekAndroidConfigurationType: ConfigurationTypeBase(
+    "Specs2RunConfiguration", // "org.spekframework.spek2-android",
```
Comment (raniejade) [inline spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidConfigurationType.kt:7]: haxx, android studio likes to hard-code strings!

File: spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidSettingsEditor.form [hunk 2]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (raniejade) [inline spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidSettingsEditor.form:1]: duplicate content... you can ignore.

File: spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekJvmRunConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import com.intellij.execution.runners.ExecutionEnvironment
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.module.ModuleManager
+import com.intellij.openapi.options.SettingsEditor
+import org.jdom.Element
+import org.spekframework.intellij.support.SpekJvmCommonRunConfigurationParameters
+import java.util.*
+
+class SpekJvmRunConfiguration(name: String, module: JavaRunConfigurationModule, factory: ConfigurationFactory)
+    : ModuleBasedConfiguration<JavaRunConfigurationModule>(name, module, factory)
```
Comment (raniejade) [inline spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekJvmRunConfiguration.kt:14]: this is the breaking change: `ModuleBasedConfiguration<JavaRunConfigurationModule>` from `ModuleBasedConfiguration`.. fun timess.

File: spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import com.intellij.execution.ui.ConsoleView
+import com.intellij.execution.util.JavaParametersUtil
+import com.intellij.openapi.extensions.Extensions
+import com.intellij.openapi.roots.OrderEnumerator
+import com.intellij.util.PathUtil
+import org.spekframework.ide.ServiceMessageAdapter
+
+class SpekRunnnerCommandLineState<T>(environment: ExecutionEnvironment, configuration: T)
+    : BaseJavaApplicationCommandLineState<T>(environment, configuration)
+where T: RunConfigurationBase,
```
Comment (raniejade) [inline spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt:24]: another one... `RunConfigurationBase` -> `RunConfigurationBase<T>` in `2018.3`

File: spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidRunConfiguration.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration> {
+        return SpekAndroidSettingsEditor(project)
+    }
+
+    override fun getState(executor: Executor, environment: ExecutionEnvironment): RunProfileState? {
+        return SpekRunnnerCommandLineState(environment, this)
+    }
+
+    override fun getValidModules(): MutableCollection<Module> {
+        return Arrays.asList(*ModuleManager.getInstance(project).modules)
```
Comment (vanniktech) [inline spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekAndroidRunConfiguration.kt]: Can't you use listOf or something so we can get rid the `*`

File: spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        JavaParametersUtil.configureModule(module, params, pathType, jreHome)
+
+        val jars = listOf(
+            PathUtil.getJarPathForClass(ServiceMessageAdapter::class.java)
+        )
+
+        params.classPath.addAll(jars)
+        params.mainClass = "org.spekframework.ide.ConsoleKt"
+
+        // this will set the vm parameters
```
Comment (vanniktech) [inline spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt]: nit: comments should be sentences. https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                .pathsList
+                .pathList
+                .forEach {
+                    params.programParametersList.add("--sourceDirs", it)
+                }
+
+        }
+
+        params.programParametersList.add("--paths", configuration.data.path.serialize())
+
```
Comment (vanniktech) [inline spek-ide-plugin/android-studio/src/main/kotlin/org/spekframework/intellij/SpekRunnnerCommandLineState.kt:68]: nit: double new line

File: spek-ide-plugin/intellij-base/src/main/kotlin/org/spekframework/intellij/support/SpekCommonProgramRunConfigurationParameters.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import com.intellij.execution.CommonProgramRunConfigurationParameters
+import com.intellij.execution.configuration.EnvironmentVariablesComponent
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.JDOMExternalizerUtil
+import org.jdom.Element
+import org.spekframework.intellij.ProducerType
+import org.spekframework.spek2.runtime.scope.Path
+import org.spekframework.spek2.runtime.scope.PathBuilder
+
+abstract class SpekCommonProgramRunConfigurationParameters(private val _project: Project): CommonProgramRunConfigurationParameters {
```
Comment (vanniktech) [inline spek-ide-plugin/intellij-base/src/main/kotlin/org/spekframework/intellij/support/SpekCommonProgramRunConfigurationParameters.kt:12]: why the underscore?

File: spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekJvmRunConfiguration.kt (hunk 3 above)
Comment (vanniktech) [reply spek-ide-plugin/intellij-idea/src/IJ182/kotlin/org/spekframework/intellij/SpekJvmRunConfiguration.kt:14]: Not!

File: spek-ide-plugin/intellij-idea/src/main/kotlin/org/spekframework/intellij/SpekJvmSettingsEditor.kt [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import com.intellij.ui.TextFieldWithHistory
+import org.spekframework.spek2.runtime.scope.PathBuilder
+import javax.swing.JPanel
+import kotlin.properties.Delegates
+
+/**
+ * @author Ranie Jade Ramiso
+ */
+class SpekJvmSettingsEditor(project: Project): SettingsEditor<SpekJvmRunConfiguration>() {
+    lateinit var panel: JPanel
```
Comment (vanniktech) [inline spek-ide-plugin/intellij-idea/src/main/kotlin/org/spekframework/intellij/SpekJvmSettingsEditor.kt]: Make them private?

**Review by vanniktech — APPROVED** [review]
I know why I've only once dipped my feet into IntelliJ Plugin development and immediately abandoned it.

> I've dropped support for Android Studio 3.1, it's easy to bring back - let me know!

3.2.0 is stable and it's been out a few weeks. So no need.

I've only commented once on those files that were duplicated.

File: spek-ide-plugin/intellij-base/src/main/kotlin/org/spekframework/intellij/support/SpekCommonProgramRunConfigurationParameters.kt (hunk 8 above)
Comment (raniejade) [reply spek-ide-plugin/intellij-base/src/main/kotlin/org/spekframework/intellij/support/SpekCommonProgramRunConfigurationParameters.kt:12]: it conflicts with `CommonProgramRunConfigurationParameters` which has a `getProject()`.

**Pull request merged**